zulipterminal/ui_tools/views.py

```diff
@@ -1593,3 +1594,81 @@ def keypress(self, size: urwid_Size, key: str) -> str:
             )
             return key
         return super().keypress(size, key)
+
+
+class EmojiPickerView(PopUpView):
+    """
+    Displays Emoji Picker view for messages.
+    """
+
+    def __init__(
+        self,
+        controller: Any,
+        title: str,
+        emoji_names: List[str],
+        message: Message,
+        view: Any,
+    ) -> None:
+        self.width = 40
+        self.view = view
+        self.message = message
+        self.controller = controller
+        emoji_buttons = self.generate_emoji_buttons(emoji_names)
+        self.emoji_search = PanelSearchBox(
+            self, "SEARCH_EMOJIS", self.update_emoji_list
+        )
+        search_box = urwid.LineBox(
+            self.emoji_search,
+            tlcorner="─",
+            tline="",
+            lline="",
+            trcorner="─",
+            blcorner="─",
+            rline="",
+            bline="─",
+            brcorner="─",
+        )
+        self.empty_search = False
+        self.allow_update_emoji_list = True
+        self.search_lock = threading.Lock()
+        super().__init__(
+            controller,
+            emoji_buttons,
+            "ADD_REACTION",
+            self.width,
+            title,
+            header=[search_box],
+        )
+
+    @asynch
+    def update_emoji_list(
+        self,
+        search_box: Any = None,
+        new_text: Optional[str] = None,
+        emoji_list: Any = None,
+    ) -> None:
+        pass
+
+    def generate_emoji_buttons(self, emoji_names: List[str]) -> List[EmojiButton]:
+        return [
+            EmojiButton(self.controller, self.width, emoji_name, self.message)
+            for emoji_name in emoji_names
+        ]
+
+    def keypress(self, size: urwid_Size, key: str) -> str:
+        if (
+            is_command_key("ENTER", key)
+            and not self.emoji_search.panel_view.controller.is_in_editor_mode()
```

Comment:
You can just use `self.controller.is_in_editor_mode()` I think.
Not related to this but I wanted to know if you knew how `enter_editor_mode` happens for SearchBox and PanelSearchBox? It is clear for WriteBox but not for these, does it happen automatically by urwid?

Reply:
You see, that's the issue.
I just noticed this yesterday after your review, that `is_in_editor_mode()` always returns `False`. So the above condition buggy and that may be the reason for the multiple reactions bug that you reported in the stream, not sure though.

It is always `False` even if I use `self.controller.is_in_editor_mode()` which makes me believe that `enter_editor_mode` never actually happens for PanelSearchBox, as you mentioned above, for any of the SearchBoxes? I can obviously explicitly `enter/exit_editor_mode` when shifting focus to/from `PanelSearchBox`, but I can't convince myself if that is a good approach. Let me know if you figure out how the editor mode works for these SearchBox, or maybe it's worth bringing this to the stream?

Reply:
But the thing is it works! Atleast in PanelSearchBox if you remove `exit_editor_mode` the search box behaves weirdly. It doesn't come out of search. That means it _was_ indeed in editor mode. But you say it's always False.

Reply:
To answer your first question. The other search boxes `enter_editor_mode` explicitly in `ui.py` since they are not part of any popups or sub-UI. So, `urwid` doesn't automatically enter editor mode, but we need to specify it explicitly. Which points that my above approach is actually what I should follow. Thanks for the discussion, I'll amend my commits and push the next iteration in a short while. :)

Reply:
Oh ok got it, forgot to check `ui.py`.
So in your case it doesn't matter if you don't `enter_editor_mode`?
BTW, `self.emoji_search.panel_view` == `EmojiPickerView` == `self`

Reply:
> So in your case it doesn't matter if you don't `enter_editor_mode`?

That's the thing that matters. Earlier, I was getting `is_in_editor_mode` always `False`, because of this reason only. So, I need to explicitly `enter_editor_mode`, as I said in my comment above, because the `PanelSearchBox` never does that.